### PR TITLE
Feature: add retry strategy

### PR DESCRIPTION
**Pull request by PabloSanchi**
# Add Flexible Retry Mechanism with Configurable Backoff and Jitter
Issue #16 

## Description:

This PR introduces a flexible retry mechanism for HTTP requests with the following key features:

- Configurable Retries: Allows setting the number of retry attempts with the WithRetries option.
- Backoff with Optional Jitter: Implements a backoff strategy with an optional jitter to prevent thundering herd problems using the WithBackoff and WithMaxJitter options.
- Custom Retry Condition: Users can define custom conditions to determine whether a retry should occur based on the error using the WithRetryIf option.
- Retry Callback: A callback function can be provided to execute custom logic after each retry attempt using the WithOnRetry option.
- Context Support: The retry process can be canceled or timed out using a context, ensuring the retries adhere to the overall request's lifecycle.

## Code Changes:

- Added a Retry function that accepts a RetryableFuncWithResponse and a variadic list of RetryOption to configure retries.
- Introduced a RetryConfig struct to hold retry configurations.
- Defined a Timer interface to abstract time-based operations for retries, allowing easier testing and customization.
- Provided several RetryOption functions to customize retry behavior (WithRetries, WithBackoff, WithMaxJitter, WithOnRetry, WithRetryIf).
- Used this lightweight retry module in the `generate` and `embedding` features.

## Tests:

Ensure to include or add unit tests to cover various retry scenarios, including:
- Success scenario (200 Code)
- Failure scenario (429 Code | too many requests)

### CODE REVIEW

**Comment by PabloSanchi** [comment]
Now it gets more flexible and it can be used with more workload avoiding the `429` error. But I will encourage to tell the user that a rate limiter in the API is important to be even more secure.

**Comment by h0rv** [comment]
Also for clarification, is jitter here the same as backoff on a failed request?

File: pkg/models/embedding.go [hunk 1]
```diff
@@ -85,28 +83,27 @@ func (m *Client) generateEmbeddingRequest(payload EmbeddingPayload) (embeddingRe
...
 	req, err := http.NewRequest(http.MethodPost, embeddingUrl, bytes.NewBuffer(payloadJSON))
+	if err != nil {
+		return embeddingResponse{}, err
+	}
 
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("Authorization", "Bearer "+m.token.value)
 
-	res, err := m.httpClient.Do(req)
+	resp, err := Retry(
```
[the remark this answers is not in the document]
Comment (PabloSanchi) [reply pkg/models/embedding.go:93]: That is a good idea, I will make a wrapper of the `http.Client` that follows the `Doer` interface. In addition, I will add DoRequestWithRetry to the interface as well.

**Comment by PabloSanchi** [comment]
The idea of the jitter is to add it to the default backoff value (which is 1 second) to prevent collisions. Each retry will now wait for `backoff + random(0, maxJitter)` seconds. [Backoff = 1 second, Jitter = 0.1 seconds] (I believe a greater jitter will be more suitable for this scenario)

To sum up, each request will have a different random value of the jitter, this helps stagger retry attempts and reduces the chances of them executing simultaneously thus preventing the too many requests `429` status code.

File: pkg/models/embedding.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (PabloSanchi) [reply pkg/models/embedding.go]: Done :)

**Review by h0rv — APPROVED** [review]
LGTM - thanks!

**Pull request merged**